libs/logstream/streamer.go

```diff
@@ -0,0 +1,506 @@
+package logstream
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+const handshakeErrorBodyLimit = 4 * 1024
+const defaultUserAgent = "databricks-cli logstream"
+
+// Dialer defines the subset of websocket.Dialer used by the streamer.
+type Dialer interface {
+	DialContext(ctx context.Context, urlStr string, requestHeader http.Header) (*websocket.Conn, *http.Response, error)
+}
+
+// TokenProvider refreshes tokens when the streamer needs a new bearer token.
+type TokenProvider func(context.Context) (string, error)
+
+// Config holds the options for running a log stream.
+type Config struct {
+	Dialer        Dialer
+	URL           string
+	Origin        string
+	Token         string
+	TokenProvider TokenProvider
+	Search        string
+	Sources       map[string]struct{}
+	Tail          int
+	Follow        bool
+	Prefetch      time.Duration
+	Writer        io.Writer
+	UserAgent     string
+}
+
+// Run connects to the log stream described by cfg and copies frames to the writer.
+func Run(ctx context.Context, cfg Config) error {
+	if cfg.Writer == nil {
+		return errors.New("logstream: writer is required")
+	}
+
+	streamer := &logStreamer{
+		dialer:        cfg.Dialer,
+		url:           cfg.URL,
+		origin:        cfg.Origin,
+		token:         cfg.Token,
+		tokenProvider: cfg.TokenProvider,
+		search:        cfg.Search,
+		sources:       cfg.Sources,
+		tail:          cfg.Tail,
+		follow:        cfg.Follow,
+		prefetch:      cfg.Prefetch,
+		writer:        cfg.Writer,
+		userAgent:     cfg.UserAgent,
+	}
+	if streamer.userAgent == "" {
+		streamer.userAgent = defaultUserAgent
+	}
+	return streamer.Run(ctx)
+}
+
+type logStreamer struct {
+	dialer        Dialer
+	url           string
+	origin        string
+	token         string
+	tokenProvider TokenProvider
+	search        string
+	sources       map[string]struct{}
+	tail          int
+	follow        bool
+	prefetch      time.Duration
+	writer        io.Writer
+	tailFlushed   bool
+	userAgent     string
+}
+
+func (s *logStreamer) Run(ctx context.Context) error {
```

Comment:
I'd add a comment that this method isn't thread-safe, just to make sure no one runs a single logStreamer instance concurrently 👍